doc/engine.md

```diff
@@ -0,0 +1,215 @@
+* Screen resolution
+* Maximum FPS
+* Vertical sync
+* Anti-aliasing
+* Texture filtering
+* Stereoscopy
+
+The above settings affect the image only in minor ways,
+and are safe to apply to just about any content.
+
+
+Examples of non-neutral features that should be API gated:
+* Direct lighting
+* Realtime shadows
+* Bump-mapping
+* FOV changes
+
+Above features affect visuals or gameplay in significant ways,
+requiring some degree of gating in the scripting API.
+It's best if the language in the config descriptions is carefully
+chosen to use words such as "allow" rather than "enable", to
+avoid confusion.
+
+It does not suffice to mark a feature as "experimental" or have an explicit
+"force" setting - assume that the user will enable everything you put
+in front of them, even to their detriment. Don't expect every user to know
+what the options actually do.
+
+A third class of visual settings would be speedhacks, such as:
+* Opaque water
+* Opaque leaves
+* No waving
+
+With speedhacks, there is a clear understanding that they are sacrificing
+quality for performance, and possibly breaking things in the process.
+In future GUI redesigns they should probably be labeled more clearly as such.
+
+Finally, texture packs are out of scope for these considerations,
+as they are a deliberate mechanism to locally customize
+the appearance of content.
+
+## 4. Upgrading
+
+Occasionally, it may be impossible to improve the engine without
+breaking compatibility in some ways. While it is never acceptable
+to break compat without providing some way to restore the original
+conditions, it may sometimes be okay to require some action from
+a mod or game author to maintain compatibility with subsequent
+engine versions. However, significant changes to existing APIs
+should only happen in a major release. Scripts should at least
+compile and run unmodified between minor releases.
+
+The standard of reasonable burden on a developer when upgrading
+to a new engine version is outlined in the following section.
+
+### 4.1. Reasonable upgrade burden
+
+The amount of work required from a game developer or modder
+to upgrade their scripts must be reasonable, and the old
+functionality must not be degraded in noticeable ways.
+Features must not consume any resources of the client
+or the server after being disabled by game code.
+
+Examples of reasonable burdens:
+* Utilizing a new API function in a few places
+* Utilizing a new object or node flag
+* Re-exporting or converting some assets (without loss of data or function)
+
+Examples of unreasonable burdens and effects:
+* Having to rewrite huge amounts of code
+* Having to send a lot more network traffic than before
+* Emulating old functionality in suboptimal ways
+* Relying on a .conf entry rather than an API function
+
+The standard of these burdens may be more relaxed in
+a major release (6.0, 7.0, etc) but under no circumstances
+may the new API heavily favor one use case over another.
+
```

Comment:
The "upgrading + compat" section is too short and doesn't touch upon the following:
* world format compatibility
* network protocol compat
  * both old client <-> new server, new server <-> old client need to work
  * where "work" obviously doesn't mean that new features magically appear on old clients

It's also way too soft on mod/game compatibility. Until now that has worked like this:
* Mods must work on new engine versions unmodified
* This only applies if the mod is coded according to the intention of the API and its documentation (hacks or using undocumented functions don't count)
* Deprecated features are removed very late (\*)
* Very few exceptions to this, pretty much only if not at all possible otherwise

(\*) Personally I prefer a grace period of 2 releases, which is e.g. what I did in the texture alpha PR and other devs seem to be fine with this.

Reply:
<b>...Unless</b> this entire section is about what is okay *when* compat is broken in which case that was not obvious to me and there needs to be a section what is okay in the normal case.

Reply:
Yeah, it's not a replacement for semver rules, but rather a clarification of what constitutes a good upgrade path.

I think I'm also being a little too soft on this and possibly confused by past experiences.
Consider #9374 and the subsequent opt-out; on one hand it was not okay to make a heavy handed change like that and ideally, if we're being responsible about these things now, it should have required action from the game code to enable. On the other, it really wasn't a lot of trouble for me to just add one flag to the CAOs so that counts as a _reasonable burden_. But maybe for someone who's less smart about CAO generation, it could require adding not three but a hundred lines of code, and that could be unreasonable. And I guess we're all paying bandwidth for the extra flag ever since.

Ideally I'd like it if visuals were treated just like behavior for these purposes. Maybe changes from opt-in to opt-out are only okay on major version steps. Then again, I'm not sure if there's any reason for ever making such a switch - making and distributing a `enable_shadows` mod for minetest_game can't be that hard, it's not impossible to even PR that into the subgame itself, and that would probably be the correct course of action for the hype fiends who want graphics in minetest; they basically play only minetest_game and its derivatives, and their pet features should not infest other projects without action.

Final remark: If we only had opt-in visual features but quietly enabled these in MTG, I don't think any of these folks would even notice and cry "evil hex isn't letting me enable shadows and gouraud wtf my FREEDOMS :o" like they usually do when it's brought up.